server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1621,22 +1619,37 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
-                FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
-                MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
-                boolean canMatch;
-                if (canRewriteToMatchNone(request.source())) {
-                    QueryBuilder queryBuilder = request.source().query();
-                    canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
-                } else {
-                    // null query means match_all
-                    canMatch = aliasFilterCanMatch;
+                if (aliasFilterCanMatch == false
+                    || (canRewriteToMatchNone(request.source()) && request.source().query() instanceof MatchNoneQueryBuilder)) {
+                    return new CanMatchResponse(false, null);
                 }
-                final FieldDoc searchAfterFieldDoc = getSearchAfterFieldDoc(request, context);
-                final Integer trackTotalHitsUpto = request.source() == null ? null : request.source().trackTotalHitsUpTo();
-                canMatch = canMatch && canMatchSearchAfter(searchAfterFieldDoc, minMax, sortBuilder, trackTotalHitsUpto);
 
-                return new CanMatchResponse(canMatch || hasRefreshPending, minMax);
+                boolean canMatch = true;
+                final FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
+                final MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
+                final Integer trackTotalHitsUpto = request.source() == null ? null : request.source().trackTotalHitsUpTo();
+                // Skipping search on shard/segment entirely can cause mismatch on total_tracking_hits, hence skip only if
+                // track_total_hits is false.
+                // Check for sort.missing == null, since in case of missing values sort queries, if segment/shard's min/max
+                // is out of search_after range, it still should be printed and hence we should not skip segment/shard.
+                if (Objects.equals(trackTotalHitsUpto, SearchContext.TRACK_TOTAL_HITS_DISABLED)
+                    && minMax != null
+                    && sortBuilder.missing() == null) {
+                    final Object primarySearchAfterField = SearchAfterBuilder.getPrimarySearchAfterFieldOrNull(request.source());
+                    if (primarySearchAfterField != null) {
+                        final FieldDoc searchAfterFieldDoc = getPrimarySearchAfterFieldDoc(sortBuilder, primarySearchAfterField, context);
+                        canMatch = canMatchSearchAfter(searchAfterFieldDoc, minMax, sortBuilder, trackTotalHitsUpto);
+                    }
+                }
+                return new CanMatchResponse(canMatch, canMatch ? minMax : null);
```

Comment:
Also, can we just not send over minMax? Should get ignored if canMatch is false anyway?

Reply:
I think that setting minMax as null is equivalent to not sending this value.

Reply:
Then we probably should not even care what we are sending for minMax if canMatch is false? So below should also work?

```
new CanMatchResponse(canMatch, minMax);
```